pkg/admissioncontroller/webhooks/auth/seed/handler.go

```diff
@@ -0,0 +1,175 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seed
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	authorizationv1beta1 "k8s.io/api/authorization/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/json"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+
+	// DecisionTimeout is the maximum time for the authorizer to take a decision. Exposed for testing.
+	DecisionTimeout = 10 * time.Second
+)
+
+func init() {
+	utilruntime.Must(authorizationv1beta1.AddToScheme(scheme))
+	utilruntime.Must(authorizationv1.AddToScheme(scheme))
+}
+
+// NewHandler creates a new HTTP handler for authorizing requests for resources related to a Seed.
+func NewHandler(log logr.Logger, authorizer authorizer.Authorizer) http.HandlerFunc {
+	h := &handler{
+		log:        log,
+		authorizer: authorizer,
+	}
+	return h.Handle
+}
+
+type handler struct {
+	log        logr.Logger
+	authorizer authorizer.Authorizer
+}
+
+func (h *handler) Handle(w http.ResponseWriter, r *http.Request) {
+	var (
+		body []byte
+		err  error
+	)
+
+	// Verify that body is non-empty
+	if r.Body == nil {
+		err = errors.New("request body is empty")
+		h.log.Error(err, "bad request")
+		h.writeResponse(w, nil, Errored(http.StatusBadRequest, err))
+		return
+	}
+
+	// Read body
+	if body, err = ioutil.ReadAll(r.Body); err != nil {
+		h.log.Error(err, "unable to read the body from the incoming request")
+		h.writeResponse(w, nil, Errored(http.StatusBadRequest, err))
+		return
+	}
+
+	// Verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		err = fmt.Errorf("contentType=%s, expected application/json", contentType)
+		h.log.Error(err, "unable to process a request with an unknown content type", "content type", contentType)
+		h.writeResponse(w, nil, Errored(http.StatusBadRequest, err))
+		return
+	}
+
+	// Both v1 and v1beta1 SubjectAccessReview types are exactly the same, so the v1beta1 type can be decoded into the
+	// v1 type. However the runtime codec's decoder guesses which type to decode into by type name if an Object's
+	// TypeMeta isn't set. By setting TypeMeta of an unregistered type to the v1 GVK, the decoder will coerce a v1beta1
+	// SubjectAccessReview to v1. The actual SubjectAccessReview GVK will be used to write a typed response in case the
+	// webhook config permits multiple versions, otherwise this response will fail.
```

Comment:
I noticed a small difference in the `SubjectAccessReviewSpec` between `v1` and `v1beta1`:

[v1beta1](https://github.com/kubernetes/api/blob/48bd8381a38a486f8b3cdf28cf7334a45b182f2e/authorization/v1beta1/types.go#L149):
```Groups []string `json:"group```

[v1](https://github.com/kubernetes/api/blob/48bd8381a38a486f8b3cdf28cf7334a45b182f2e/authorization/v1/types.go#L140):
```Groups []string `json:"groups```

So depending on what version is sent by the API server, I suspect that we lose the group information during decode,  right?

Reply:
👀 nice catch, I will check it

Reply:
Hi @timuthy, I've addressed this in 031f21a, please check it out. I'm not too happy with it, but I didn't have better ideas. Maybe you have one?